samples/dnn/gpt2_inference.py

```diff
@@ -0,0 +1,108 @@
+'''
+This is a sample script to run GPT-2 inference in OpenCV using ONNX model.
+The script loads the GPT-2 model and runs inference on a given prompt.
+Currently script only works with fixed size window, that means
+you will have to specify prompt of the same length as when model was exported to ONNX.
+
+
+Exporting GPT-2 model to ONNX.
+To export GPT-2 model to ONNX, you can use the following procedure:
+
+1. Clone fork of Andrej Karpathy's GPT-2 repository:
+
+    git clone https://github.com/Abdurrahheem/build-nanogpt/tree/ash/export-gpt2-onnx
+
+2. Install the required dependencies:
+
+    pip install -r requirements.txt
+
+3  Export the model to ONNX:
+
+    python export2onnx.py --promt=<Any-promt-you-want> --batch_size=<batch-size>
+
+
+Run the script:
+
+    python gpt2_inference.py --model=<path-to-onnx-model> --max_seq_len=<max-output-lenght> --batch_size=<use-one-used-while-exportinh> --prompt=<use-promt-of-the-same-length-used-while-exporting>
+'''
+
+
+
+from copy import deepcopy
+import numpy as np
+import tiktoken
+import argparse
+import cv2
+
+def parse_args():
+    parser = argparse.ArgumentParser(description='Use this script to run GPT-2 inference in OpenCV',
+                                    formatter_class=argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--model', type=str, required=True, help='Path to GPT-2 model ONNX model file.')
+    parser.add_argument("--max_seq_len", type=int, default=30, help="Number of tokens to continue.")
+    parser.add_argument("--batch_size", type=int, default=5, help="Number of batches.")
+    parser.add_argument("--prompt", type=str, default="Hello, I'm a language model,", help="Prompt to start with.")
+    return parser.parse_args()
+
+def stable_softmax(logits):
+    exp_logits = np.exp(logits - np.max(logits, axis=-1, keepdims=True))
+    return exp_logits / np.sum(exp_logits, axis=-1, keepdims=True)
+
+
+def gpt2_inference(net, tokens, max_length, num_return_sequences=5):
+
+    print("Inferencing GPT-2 model...")
+    x = np.array(tokens)
+    x = np.tile(x, (num_return_sequences, 1))
+
+    output_buffer = deepcopy(x)
+    counter = x.shape[1]
+    while counter < max_length:
+
+        net.setInput(x)
+        logits = net.forward()
+
+        # logits is assumed to be (B, seq_length, vocab_size) and needs to be the last token's logits
+        logits = logits[:, -1, :]  # (B, vocab_size)
+
+        # Get the probabilities using softmax
+        probs = stable_softmax(logits)
+
+        # Do top-k sampling of 50
+        topk_indices = np.argpartition(probs, -50, axis=-1)[:, -50:]
+        topk_probs = np.take_along_axis(probs, topk_indices, axis=-1)
+
+        # Normalize top-k probabilities
+        topk_probs /= np.sum(topk_probs, axis=-1, keepdims=True)
```

Comment:
Do we really need softmax and normalization to just get top-k predictions? I think their order is not changed from raw logits.

Reply:
`np.random.choice` expect probabilities. otherwise you are correct